File: scripts/node.py
from enum import Enum
import numpy as np
from math import sin, cos, pi
from vector import Vector


class DirState(Enum):
    FORWARD = 1
    BACKWARD = -1


class Node:
    ID = 0
    MARGIN_MAX = 0.25
    MARGIN_MIN = 0.15
    MAX_ANGLE = pi / 10
    NEIGHBOUR_MAX = MARGIN_MAX + .5

    def __init__(self, pos, angle, cost, dir=DirState.FORWARD, parent=None):
        # type: (Vector, float, float, DirState, Node| None) -> None
        self.pos = pos
        self.dir = dir
        self.angle = angle
        self.parent = parent
        self.childs = []  # type: list[Node]
        self.id = Node.ID
        Node.ID += 1
        self.cost = cost

    @property
    def direction(self):
        # type: () -> Vector
        x = cos(self.angle)
        y = sin(self.angle)
        return Vector(x, y, 0)

    def __eq__(self, other):
        # type: (object) -> bool
        if not isinstance(other, Node):
            return False

        return other.id == self.id

    @classmethod
    def refactor_length(cls, length):
        # type: (float) -> float
        if length > cls.MARGIN_MAX:
            return cls.MARGIN_MAX
        elif length < cls.MARGIN_MIN:
            return cls.MARGIN_MIN
        else:
            return length

    def can_be_child(self, other):
        # type: (Node) -> bool
        direction = other.pos - self.pos
        if direction.length > self.NEIGHBOUR_MAX:
            return False
        angle = self.direction.angle_to(direction)
        angle_abs = abs(angle)
        if angle_abs > 90 and not 180 - angle_abs > 180 - self.MARGIN_MAX / 2:
            return False
        elif not angle_abs < self.MAX_ANGLE / 2:
            return False

        new_angle = self.get_angle_from_child(other.pos)
        for child in other.childs:
            dif = abs(child.angle - new_angle)
            if dif > self.MAX_ANGLE / 2:
                return False

        return True

    def is_in_range(self, point):
        # type: (Vector) -> tuple[bool, bool]
        direction = point - self.pos
        if direction.length > self.NEIGHBOUR_MAX:
            return False, False
        angle = abs(self.direction.angle_to(direction))
        if angle > 90:
            return 180 - angle > 180 - self.MARGIN_MAX / 2, True
        else:
            return angle < self.MAX_ANGLE / 2, False

    def is_close_enough(self, goal_pos):
        # type: (Vector) -> bool
        return (self.pos - goal_pos).length < self.MARGIN_MAX

    def get_angle_from_child(self, child):
        # type: (Vector) -> float
        direction = child - self.pos
        return np.arctan2(direction.y, direction.x)  # type: ignore

    def add_child(self, pos, dir_changed):
        # type: (Vector, bool) -> Node
        angle = self.get_angle_from_child(pos)
        if self.dir is DirState.FORWARD:
            if dir_changed:
                dir = DirState.BACKWARD
            else:
                dir = DirState.FORWARD
        else:
            if dir_changed:
                dir = DirState.FORWARD
            else:
                dir = DirState.BACKWARD

        cost = self.cost + (self.pos - pos).length
        child = Node(pos, angle, cost, dir, self)
        self.childs.append(child)
        return child

    def update_childs_cost(self):
        for child in self.childs:
            new_cost = (self.pos - child.pos).length
            child.cost = self.cost + new_cost
            child.update_childs_cost()

    def update_parent(self, new_parent, new_cost):
        # type: (Node, float) -> None
        if self.parent:
            self.parent.childs.remove(self)

        self.parent = new_parent
        self.parent.childs.append(self)

        self.angle = self.parent.get_angle_from_child(self.pos)
        self.cost = new_cost
        self.update_childs_cost()

File: src/node.cpp
#include "node.hpp"
#include <cmath>
#include <vector>
#include <algorithm>

// static elements initialize
int Node::ID = 0;

Node::Node(Vector pos, float angle, float cost, DirState dir, Node *parent = nullptr) : pos{pos}, m_dir{dir}, angle{angle}, parent{parent}, cost{cost}
{
    m_id = ID;
    ID++;
    childs = std::vector<Node *>();
}

float Node::get_angle_from_child(const Vector child) const
{
    return (child - pos).angle();
}

Vector Node::direction() const
{
    return Vector{cos(angle), sin(angle), 0};
}

float Node::refactor_angle(float length) const
{
    if (length > MARGIN_MAX)
    {
        return MARGIN_MAX;
    }
    if (length < MARGIN_MIN)
    {
        return MARGIN_MIN;
    }
    return length;
}

bool Node::can_be_child(const Node *other) const
{
    Vector direction = other->pos - pos;
    if (direction.length() > NEIHGBOUR_MAX)
    {
        return false;
    }
    float angle = this->direction().angle_to(direction);
    float angle_abs = abs(angle);

    if (angle_abs > M_PI && !(M_PI * 2 - angle_abs > M_PI * 2 - MARGIN_MAX / 2))
    {
        return false;
    }

    if (!angle_abs < MAX_ANGLE / 2)
    {
        return false;
    }

    float new_angle = get_angle_from_child(other->pos);
    for (auto child : other->childs)
    {
        float dif = abs(child->angle - new_angle);
        if (dif > MAX_ANGLE / 2)
            return false;
    }

    return true;
}

ReversedState<bool> Node::is_in_range(const Vector point) const
{
    Vector direction = point - pos;
    if (direction.length() > NEIHGBOUR_MAX)
        return ReversedState<bool>{false, false};

    float angle = abs(this->direction().angle_to(direction));

    if (angle > M_PI)
    {
        bool is_in_range = M_PI * 2 - angle > M_PI * 2 - MARGIN_MAX / 2;
        return ReversedState<bool>{is_in_range, true};
    }

    return ReversedState<bool>{angle < MAX_ANGLE / 2, false};
}

bool Node::is_close_enough(const Vector goal_pos) const
{
    return (pos - goal_pos).length() < MARGIN_MAX;
}

void Node::update_childs_cost()
{
    for (auto child : childs)
    {
        float new_cost = (pos - child->pos).length();
        child->cost = cost + new_cost;
        child->update_childs_cost();
    }
}

void Node::update_parent(Node *new_parent, float new_cost)
{
    if (parent)
    {
        auto it = std::find(parent->childs.begin(), parent->childs.end(), this);
        if (it != parent->childs.end())
            parent->childs.erase(it);
    }

    parent = new_parent;
    parent->childs.push_back(this);

    angle = parent->get_angle_from_child(pos);
    cost = new_cost;
    update_childs_cost();
}